### PR TITLE
creates a snackbar that appears when app messages are recorded

### PR DESCRIPTION
**Pull request by hamilton**
This PR creates a new app message notification system using `material-ui` snackbars. This will be useful for all sorts of app-level events, like saving to local storage, exporting to gist, and the like.

It's not particularly useful right now, and it users the probably-too-simplistic format of just an array of strings (so no time in which the app message happened, level, understanding of componentizing different types of messages, etc.)

I think one review would be enough, if any of y'all could take a look.

### CODE REVIEW

File: src/actions/task-definitions.js [hunk 1]
```diff
@@ -7,9 +7,7 @@ import {
   viewModeIsEditor,
   getCells,
   getCellBelowSelectedId,
-  getCellAboveSelectedId,
-  prettyDate,
-  formatDateString,
+  getCellAboveSelectedId, prettyDate, formatDateString,
```
Comment (bcolloran) [inline src/actions/task-definitions.js]: why this change?

File: src/actions/actions.js [hunk 2]
```diff
@@ -43,10 +43,13 @@ export function importNotebook(newState) {
 export function importFromURL(importedState) {
   return (dispatch) => {
     dispatch(importNotebook(importedState))
+    dispatch(updateAppMessages('Notebook successfully imported from URL.'))
     return Promise.resolve()
   }
 }
 
+//  dispatch(updateAppMessages(`gonna run all cells. ${(new Date()).toString()}`))
```
Comment (bcolloran) [inline src/actions/actions.js]: can cut this comment

File: src/actions/task-definitions.js (hunk 1 above)
Comment (hamilton) [reply src/actions/task-definitions.js]: Oh I think that was a weird linting thing tbh

**Comment by hamilton** [comment]
@bcolloran re: accepting objects, gladly. I had thought it was preferred to stick to strings before, but this is clearly a more worthwhile paradigm.

**Comment by bcolloran** [comment]
ah, gotcha -- to clarify: it is preferred to stick with objects that can be serialized simply (so e.g. no complicated objects with methods, like e.g. date objects) and ideally objects that have a fixed schema (but this is not a requirement). but an object containing a bunch of strings, ints, bools is not a problem at all. and if the fields are fixed and enumerable, all the better. For now, i think an appMessage schema along the lines of `{"message": "string", "details": "string"}` (and perhaps "timestamp":"string" if you feel like it) should cover most of our needs.

i'm thinking of the gist export use case (which will be one of our first), wherein we know we'll want an app message with a URL, which won't fit in the snackbar. for now if we allow the "details" str to be an html snippet, and we dangerourslySetInnerHtml, we can achieve that (as is the case currently in app-info-pane.jsx -- i think there is one line in that file that would need to be changed)

**Comment by bcolloran** [comment]
just noticed that i'm getting the following in the console when i load a url with the gist trick, eg
`/iodide/dev/iodide.dev.html?url=https://iodide-project.github.io/iodide-examples/output-handling.html` --

```
Warning: React does not recognize the `SnackbarContentProps` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `snackbarcontentprops` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
    in div (created by Snackbar)
    in EventListener (created by ClickAwayListener)
    in ClickAwayListener (created by Snackbar)
    in Snackbar (created by WithStyles(Snackbar))
    in WithStyles(Snackbar) (created by appMessagesUnconnected)
    in appMessagesUnconnected (created by Connect(appMessagesUnconnected))
    in Connect(appMessagesUnconnected) (created by Page)
    in div (created by Page)
    in Page (created by Connect(Page))
    in Connect(Page)
    in Provider
```

**Comment by hamilton** [comment]
Great - thanks for the heads up!

File: src/components/app-messages/app-messages.jsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    if (latestMessage !== newMessage) this.setState({ latestMessage: newMessage, open: true })
+    return true
+  }
+
+  handleClose = () => {
+    this.setState({ open: false });
+  }
+
+  handleMore = () => {
+    console.log('all right!')
```
Comment (wlach) [inline src/components/app-messages/app-messages.jsx]: probably don't need this console.log :)

**Comment by hamilton** [comment]
I'm not crazy about using `dangerouslySetHtml`, since it feels like we're punting on actually making useful log reps here that are testable and reasonable.

**Comment by bcolloran** [comment]
just noticed that i'm getting the following in the console when i load a url with the gist trick, eg
`/iodide/dev/iodide.dev.html?url=https://iodide-project.github.io/iodide-examples/output-handling.html` --

```
Warning: React does not recognize the `SnackbarContentProps` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `snackbarcontentprops` instead. If you accidentally passed it from a parent component, remove it from the DOM element.
    in div (created by Snackbar)
    in EventListener (created by ClickAwayListener)
    in ClickAwayListener (created by Snackbar)
    in Snackbar (created by WithStyles(Snackbar))
    in WithStyles(Snackbar) (created by appMessagesUnconnected)
    in appMessagesUnconnected (created by Connect(appMessagesUnconnected))
    in Connect(appMessagesUnconnected) (created by Page)
    in div (created by Page)
    in Page (created by Connect(Page))
    in Connect(Page)
    in Provider
```

**Comment by hamilton** [comment]
Chatted w/ @bcolloran about the dangerous inner html issue, and we've agreed that the right direction for this will be to create app message components, but for now we're more interested in getting this together as-is.

**Comment by hamilton** [comment]
@bcolloran snackbar PR is ready for a new review. Updated schema for app messages to be an object, ordered entries by recency, added time stamp to item which now is in the pane.

**Comment by bcolloran** [comment]
hmm, i'm getting:

```
handle-url-query.js:18 failed to load notebook url TypeError: Cannot read property 'when' of undefined
    at appMessagesUnconnected.shouldComponentUpdate (app-messages.jsx:25)
    at checkShouldComponentUpdate (react-dom.development.js:6696)
    at updateClassInstance (react-dom.development.js:7208)
    at updateClassComponent (react-dom.development.js:8345)
    at beginWork (react-dom.development.js:8982)
    at performUnitOfWork (react-dom.development.js:11814)
    at workLoop (react-dom.development.js:11843)
    at renderRoot (react-dom.development.js:11874)
    at performWorkOnRoot (react-dom.development.js:12449)
    at performWork (react-dom.development.js:12370)
```

when i visit `iodide/dev/output-handling.html?url=https://iodide-project.github.io/iodide-examples/output-handling.html`

File: src/actions/actions.js (hunk 2 above)
Comment (bcolloran) [reply src/actions/actions.js]: do you wish to retain this comment?

File: src/actions/task-definitions.js (hunk 1 above)
Comment (bcolloran) [reply src/actions/task-definitions.js]: i have been hearing such good things about "prettier". we should consider using that so that we never have to think/talk about boring linting stuff like evar. :-)

**Comment by bcolloran** [comment]
hmmm, actually i get that error if i just open a blank notebook `iodide/dev/iodide.dev.html`, and without doing anything else, if i click any of the buttons to open any of the side panes, i get
```
app-messages.jsx:25 Uncaught TypeError: Cannot read property 'when' of undefined
    at appMessagesUnconnected.shouldComponentUpdate (app-messages.jsx:25)
    at checkShouldComponentUpdate (react-dom.development.js:6696)
    at updateClassInstance (react-dom.development.js:7208)
    at updateClassComponent (react-dom.development.js:8345)
    at beginWork (react-dom.development.js:8982)
    at performUnitOfWork (react-dom.development.js:11814)
    at workLoop (react-dom.development.js:11843)
    at HTMLUnknownElement.callCallback (react-dom.development.js:100)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:138)
    at invokeGuardedCallback (react-dom.development.js:187)
shouldComponentUpdate @ app-messages.jsx:25
checkShouldComponentUpdate @ react-dom.development.js:6696
updateClassInstance @ react-dom.development.js:7208
updateClassComponent @ react-dom.development.js:8345
beginWork @ react-dom.development.js:8982
performUnitOfWork @ react-dom.development.js:11814
workLoop @ react-dom.development.js:11843
callCallback @ react-dom.development.js:100
invokeGuardedCallbackDev @ react-dom.development.js:138
invokeGuardedCallback @ react-dom.development.js:187
replayUnitOfWork @ react-dom.development.js:11318
renderRoot @ react-dom.development.js:11885
performWorkOnRoot @ react-dom.development.js:12449
performWork @ react-dom.development.js:12370
performSyncWork @ react-dom.development.js:12347
interactiveUpdates @ react-dom.development.js:12597
interactiveUpdates @ react-dom.development.js:1958
dispatchInteractiveEvent @ react-dom.development.js:4259


react-dom.development.js:9643 The above error occurred in the <appMessagesUnconnected> component:
    in appMessagesUnconnected (created by Connect(appMessagesUnconnected))
    in Connect(appMessagesUnconnected) (created by Page)
    in div (created by Page)
    in Page (created by Connect(Page))
    in Connect(Page)
    in Provider

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.development.js:9643
logError @ react-dom.development.js:9682
commitErrorLogging @ react-dom.development.js:9895
commitAllLifeCycles @ react-dom.development.js:11459
callCallback @ react-dom.development.js:100
invokeGuardedCallbackDev @ react-dom.development.js:138
invokeGuardedCallback @ react-dom.development.js:187
commitRoot @ react-dom.development.js:11594
completeRoot @ react-dom.development.js:12502
performWorkOnRoot @ react-dom.development.js:12452
performWork @ react-dom.development.js:12370
performSyncWork @ react-dom.development.js:12347
interactiveUpdates @ react-dom.development.js:12597
interactiveUpdates @ react-dom.development.js:1958
dispatchInteractiveEvent @ react-dom.development.js:4259



app-messages.jsx:25 Uncaught TypeError: Cannot read property 'when' of undefined
    at appMessagesUnconnected.shouldComponentUpdate (app-messages.jsx:25)
    at checkShouldComponentUpdate (react-dom.development.js:6696)
    at updateClassInstance (react-dom.development.js:7208)
    at updateClassComponent (react-dom.development.js:8345)
    at beginWork (react-dom.development.js:8982)
    at performUnitOfWork (react-dom.development.js:11814)
    at workLoop (react-dom.development.js:11843)
    at renderRoot (react-dom.development.js:11874)
    at performWorkOnRoot (react-dom.development.js:12449)
    at performWork (react-dom.development.js:12370)
```

**Comment by hamilton** [comment]
hmm, that's quite odd - wasn't getting those errors before. Thanks for the review - will check now!

File: src/actions/actions.js (hunk 2 above)
Comment (hamilton) [reply src/actions/actions.js]: meant to cut this before, apologies / thanks for catching again

**Comment by hamilton** [comment]
@bcolloran addressed the above issues. Thanks for the additional review! Clearly needed.

**Comment by bcolloran** [comment]
screwing around with an edge case, but one that will for sure come up -- firing a bunch of app messages in short succession (for example, in the case of multiple JSMD parsing errors at notebook loading time) -- i get the following warning.

```
Warning: Encountered two children with the same key, `msg-Wed May 30 2018 11:30:04 GMT-0700 (PDT)`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
    in div (created by Resizable)
    in Resizable (created by SidePaneUnconnected)
    in div (created by Paper)
    in Paper (created by WithStyles(Paper))
    in WithStyles(Paper) (created by Drawer)
    in Transition (created by Slide)
    in EventListener (created by Slide)
    in Slide (created by WithTheme(Slide))
    in WithTheme(Slide) (created by Drawer)
    in div (created by Drawer)
    in Drawer (created by WithStyles(Drawer))
    in WithStyles(Drawer) (created by SidePaneUnconnected)
    in MuiThemeProvider (created by SidePaneUnconnected)
    in SidePaneUnconnected (created by Connect(SidePaneUnconnected))
    in Connect(SidePaneUnconnected) (created by AppInfoPaneUnconnected)
    in AppInfoPaneUnconnected (created by Connect(AppInfoPaneUnconnected))
    in Connect(AppInfoPaneUnconnected) (created by ViewControls)
    in div (created by ViewControls)
    in ViewControls (created by EditorModeToolbarUnconnected)
    in div (created by Toolbar)
    in Toolbar (created by WithStyles(Toolbar))
    in WithStyles(Toolbar) (created by EditorModeToolbarUnconnected)
    in div (created by EditorModeToolbarUnconnected)
    in EditorModeToolbarUnconnected (created by Connect(EditorModeToolbarUnconnected))
    in Connect(EditorModeToolbarUnconnected) (created by NotebookHeader)
    in MuiThemeProvider (created by NotebookHeader)
    in div (created by NotebookHeader)
    in NotebookHeader (created by Page)
    in div (created by Page)
    in Page (created by Connect(Page))
    in Connect(Page)
    in Provider
```

not going to block on this, but it's an issue we'll need to fix (should be as simple as incrementing a global app message counter in the store) 

I'll let you make the call on whether you want to tackle this now since you're spun up on this code ATM, or whether you favor just merging this and dealing with it later (not sure about pros/cons of the value of getting this in and being able to use it vs the cost of a touch of technical debt in this area)

oh, but if you do decide to merge at this time, please file an issue so we don't lose this

**Comment by hamilton** [comment]
This should be fixable by switching to a Unix timestamp in raw ms, so I’ll
take care of this in this pr

**Comment by hamilton** [comment]
@bcolloran went ahead and added ids to the `appMessage`for keying etc.

File: src/state-prototypes.js [hunk 4]
```diff
@@ -256,6 +258,22 @@ function addNewCellToState(state, cellType = 'code', language = 'js') {
...
+  }
+}
+
+function addAppMessageToState(state, appMessage) {
+  const nextAppMessageId = newCellID(state.appMessages)
+  state.appMessages
+    .push(newAppMessage(nextAppMessageId, appMessage.message, appMessage.details, appMessage.when))
+  return state
+}
+
```
Comment (bcolloran) [inline src/state-prototypes.js]: i think this function should live in a `notebook-reducer-utils.js` file rather than in the state prototypes

File: src/reducers/notebook-reducer.js [hunk 5]
```diff
@@ -46,13 +46,13 @@ function clearHistory(loadedState) {
 function updateAppMessages(state, messageObj) {
   const message = Object.assign({}, messageObj)
   delete message.type
-  const appMessages = state.appMessages.slice()
   if (message.when === undefined) {
     message.when = (new Date()).toString()
```
Comment (bcolloran) [inline src/reducers/notebook-reducer.js]: if the `message.when` is required by the message schema, this should never be undefined right? is this check needed?

File: src/state-prototypes.js (hunk 4 above)
Comment (bcolloran) [reply src/state-prototypes.js]: or it could go under `updateAppMessages` directly in notebook-reducer (since we already have a few helper functions there)

File: src/reducers/notebook-reducer.js [hunk 6]
```diff
@@ -43,6 +43,18 @@ function clearHistory(loadedState) {
   /* eslint-enable */
 }
 
+function updateAppMessages(state, messageObj) {
+  const message = Object.assign({}, messageObj)
+  delete message.type
```
Comment (bcolloran) [inline src/reducers/notebook-reducer.js]: `message.type` is not in the schema and doesn't appear to exist anywhere else. can this be deleted?

File: src/reducers/notebook-reducer.js [hunk 7]
```diff
@@ -43,6 +43,18 @@ function clearHistory(loadedState) {
...
+  if (message.when === undefined) {
+    message.when = (new Date()).toString()
+  }
+  if (message.details === undefined) message.details = message.message
+  const nextState = Object.assign({}, state)
+  nextState.appMessages = nextState.appMessages.slice()
+  return addAppMessageToState(nextState, message)
+}
```
Comment (bcolloran) [inline src/reducers/notebook-reducer.js]: these checks for message.when and message.details should be part of the message constructor rather than the reducer.

File: src/state-prototypes.js (hunk 4 above)
Comment (hamilton) [reply src/state-prototypes.js]: Yeah, `addNewCellToState` is also in this file, so I figured it should go here as well. Happy to change it if you think it belongs elsewhere.

File: src/reducers/notebook-reducer.js (hunk 6 above)
Comment (hamilton) [reply src/reducers/notebook-reducer.js]: sort of ... I think I need to change how the action dispatches so I don't have to do this. This is here because I'm loading an action (which has a `type`) with a all the message attributes.

File: src/reducers/notebook-reducer.js (hunk 5 above)
Comment (hamilton) [reply src/reducers/notebook-reducer.js]: `message.when` is required by the message schema, but that doesn't mean a message will at this point have the correct schema. I believe that only happens after the reducer has completed. I suppose I should just add this into the action as the default.

**Comment by hamilton** [comment]
Thanks for the review @bcolloran! Made those changes, with the exception of the `addAppMessageToState` moving out of `state-prototypes.js`. Totally happy to do it, but I put it there originally because `addCellToState` is there as well, so let me know how you'd like to move forward there.

**Comment by bcolloran** [comment]
i think `addCellToState` is legacy and should be moved also (actually, it can be refactored out, see #642) -- probably not the code organization template we want to follow. i think it makes most sense to either include these state manipulation functions at the top of the reducer or in a reducer-utils file. i'm r+ pending that change

**Comment by hamilton** [comment]
Cool - that makes sense. I am not too familiar with that part of the code base so I defer to you here :). will migrate that over to the reducer, double-check my tests, then merge.
